rust/src/table/convert_to_delta.rs

```diff
@@ -0,0 +1,765 @@
+    /// Arrow error
+    #[error("Arrow error: {0}")]
+    Arrow(#[from] ArrowError),
+    /// Parquet error
+    #[error("Parquet error: {0}")]
+    Parquet(#[from] ParquetError),
+    /// DeltaTable error
+    #[error("DeltaTable error: {0}")]
+    DeltaTable(#[from] DeltaTableError),
+    /// Error percent-decoding as UTF-8
+    #[error("Error percent-decoding as UTF-8: {0}")]
+    PercentDecode(#[from] Utf8Error),
+    /// Error converting usize to i64
+    #[error("Error converting usize to i64: {0}")]
+    TryFromUsize(#[from] TryFromIntError),
+    /// Error reaching the end of an iterator
+    #[error("Error reaching the end of an iterator")]
+    EndOfIterator,
+    /// No parquet file is found in the given path
+    #[error("No parquet file is found in the given path")]
+    ParquetFileNotFound,
+    /// Partition column provided by the user does not exist in the parquet files
+    #[error("Partition column provided by the user does not exist in the parquet files")]
+    PartitionColumnNotExist(HashSet<SchemaField>),
+}
+
+/// Convert a Parquet table to a Delta table in place
+///
+/// # Arguments
+///
+/// * `storage` - A shared reference to a [`DeltaObjectStore`] with "/" pointing at delta table root (i.e. where `_delta_log` is located).
+/// * `partition_schema`
+///     - A vector of [`SchemaField`] describing the data type, nullibility and metadata of partion columns.
+///     - If the schema of a partition column is not provided, schema data type is set to string and the field is set to nullable.
+///     - Errors if any partition column provided by the user does not exist in the parquet files.
+pub async fn convert_to_delta(
+    storage: ObjectStoreRef,
+    partition_schema: Vec<SchemaField>,
+) -> Result<DeltaTable, Error> {
+    if storage.is_delta_table_location().await? {
+        info!("A Delta table already exists in the given object store");
+        let mut table = DeltaTable::new(storage, DeltaTableConfig::default());
+        table.load().await?;
+        return Ok(table);
+    }
+    info!(
+        "Converting Parquet table in object store location: {:?}",
+        storage.root_uri()
+    );
+    let parquet_files = parquet_files(storage.clone()).await?;
+    delta_table(parquet_files, storage, partition_schema).await
+}
+
+/// Convert a Parquet table to a Delta table in place
+///
+/// # Arguments
+///
+/// * `path` - A path with "/" pointing at delta table root (i.e. where `_delta_log` is located).
+/// * `partition_schema`
+///     - A vector of [`SchemaField`] describing the data type, nullibility and metadata of partion columns.
+///     - If the schema of a partition column is not provided, schema data type is set to string and the field is set to nullable.
+///     - Errors if any partition column provided by the user does not exist in the parquet files.
+pub async fn convert_to_delta_from_path(
+    path: &str,
+    partition_schema: Vec<SchemaField>,
+) -> Result<DeltaTable, Error> {
```

Comment:
We should provide a code path where we pass through storage options, otherwise this wont work for any remote store, or rather force config via the env.

Reply:
Done.